design/client-policies.md

```diff
@@ -0,0 +1,191 @@
+
+### 3. Conformance to a profile such as FAPI
+
+It is treated in [Client Conformance Profiles](https://issues.redhat.com/browse/KEYCLOAK-11612) based on this Client Policies. Which kind of security profiles are supported and how to realize them are discussed on 
+[Client Conformance Profiles](https://issues.redhat.com/browse/KEYCLOAK-11612).
+
+## Implementation details
+
+Client Policies consists of the three building blocks, Policy, Condition, and Executor.
+
+### Condition : which client
+
+A condition determins to which client a policy is adopted. The condition checks whether one specified criteria is satisfied. Such the task is that checking whether the [access type](https://www.keycloak.org/docs/latest/server_admin/index.html#oidc-clients) of the client is confidential.
+
+#### How to implement
+
+A condition can be implemented as a provider (`ClientPolicyConditionSpi`, `ClientPolicyConditionFactory`, `ClientPolicyCondition`). Developers can add their own condition by implementing this provider.
+
+#### How to configure
+
+A condition can be configuable the same as other configurable provider. What can be configured depends on each condition's nature.
+
+#### Which kind of conditions are provided
+
+The following conditions are provided:
+
+* The way of creating/updating a client (OIDC Dynamic Client Registration - Authenticated, Anonymous, Admin REST API - Admin Console, etc.)
+
+E.g. When creating a client, a policy is adopted when this client is created by OIDC Dynamic Client Registration with Initial Access Token.
+
+* Client Access Type (confidential, public, bearer-only)
+
+E.g. When a client sends an authorization request, a policy is adopted if this client is confidential.
+
+* Client ID
+
+E.g. When a client sends a token request, a policy is adopted if this client's ID (`client_id` in OAuth2) is "TPP-ClientApp".
+
+* Client Redirect URI
+
+E.g. When updating a client on Admin Console, a policy is adopted if this client's updating redirect URI is exactly matched with "https://www.example.com/callback/".
+
+The followings are adpoted only when creating or updating clients:
+
+* Client Role
+* Client Scope
+* Group
+* Host
+* IP
+* User
+* User Role
+
+These conditions are evaluated against some properties or contexts of a client application(e.g. `security-admin-console`) and its user who is trying to create or update clients.
+
+E.g. When creating a client, a policy is adopted if it is on Admin Console and the user of this Admin Console belongs to the group "IT-Dept".
+
+### Executor : what action
+
+An executor specifies what action is executed on a client to which a policy is adopted. The executor executes one specified action. Such the action is checking whether the value of the parameter "redirect_uri" in the authorization request matches exactly with one of the pre-registered redirect URIs on Authorization Endpoint and rejecting this request if not. 
```

Comment:
Would be good to mention what "events" an executor can act on, and that an executor can take actions on one or more events. We should also allow executors that make changes to the config.

I can think of the following events:

1. Create client - augment phase  - an executor can modify the config in this phase
2. Create client - validation phase
3. Update client - augment phase - an executor can modify the config in this phase
4. Update client - validation phase
5. Request (multiple events here I guess?)

In 5 not sure if that's all, or if there are more nuanced events here.

We probably also need executors to be protocol specific.

Reply:
I think "events" you've mentioned are nearly equal to what I've wrote in "Endpoint and Operation" section.
Incorporating your points, I'll add descriptions about "events" in "Executor" section.

Also, I think "modify the config" you've mentioned means that the executor can modify the config of the client when creating/updating client (e.g. the author of creating the client specify `token_endpoint_auth_method=client_secret_basic` but the executor change it to `privaty_key_jwt`). Is that right?

Reply:
Yes, not sure that's the best example, was more thinking along the lines of enabling/requiring consent. Enabling/requiring PKCE. However, there shouldn't be any limit to what it can do. Also, that means there's two phases in creation/updating. As otherwise "validation" may not be the finished config, but some interim.

Reply:
Thanks. As you mentioned, I'll also add multi-phase action of executor in creation/updating client to "Executor" section.